lib/netext/transport.go

```diff
@@ -0,0 +1,229 @@
+package netext
+
+import (
+	"bytes"
+	"encoding/base64"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/ThomsonReutersEikon/go-ntlm/ntlm"
+	"github.com/loadimpact/k6/lib"
+	"github.com/loadimpact/k6/stats"
+	"github.com/pkg/errors"
+)
+
+type Transport struct {
+	*http.Transport
+	options   lib.Options
+	tags      map[string]string
+	trail     *Trail
+	tlsInfo   TLSInfo
+	samplesCh chan<- stats.SampleContainer
+
+	// ntlm authentication
+	mu              sync.Mutex
+	ntlmAuthCache   map[string]bool
+	enableNtlmCache bool
+}
+
+func NewTransport(transport *http.Transport, samplesCh chan<- stats.SampleContainer, options lib.Options, tags map[string]string) *Transport {
+	return &Transport{
+		Transport:       transport,
+		tags:            tags,
+		options:         options,
+		samplesCh:       samplesCh,
+		ntlmAuthCache:   make(map[string]bool),
```

Comment:
Wasn't this supposed to be persisted across the while connection? IIRC, there was something about NTLM being authenticated per connection, while with the current code this cache will live only for a single HTTP request (whether that's a single `RoundTrip()` or a chain of redirects). Seems to me like the NTLM handling and the new per-redirect metrics should be on two different levels, the NTLM one being a per-VU or a per-iteration thing, while the redirect metric tracking being on a per-request basis.

Reply:
Yeah, I vaguely recollect the NTLM auth being per-connection as well (and reading the comments in PR https://github.com/loadimpact/k6/pull/556 seems to confirm it). That'd mean we should probably keep the NTLM auth cache at a higher level (like in `common/state.go`), but then I suppose we can't just use the URL as the cache key, but rather something that identifies the TCP connection in question right?

Reply:
Hmm not sure we can identify the TCP connection, since it would be unknown at this time and hidden by the underlying `http.Transport` storage...

Thinking about it again, why do we actually need that "cache"? It doesn't actually cache anything, it just [tries to determine if it should first send the `Authorization` header or not](https://github.com/loadimpact/k6/blob/983c2debfff728ba301348a797adb2de14c85681/lib/netext/transport.go#L149-L162)... And regardless of whether we send it or not, if we receive 401, we try to authenticate. But if we remove the cache altogether and first try the request without authentication, then authenticate when necessary, won't it still work? The only part I'm not sure about is [this magic string](https://github.com/loadimpact/k6/blob/fix/trace-redirect-requests/lib/netext/transport.go#L153), which I can't find anywhere else with google...

Reply:
You are right, I moved the code to keep in a single place but I forgot the authentications are per connection.

Regarding the "cache" it is just a naming problem, I couldn't find a better name, but I would be happy to chance once you suggest a better one.

But we need it because once the connection is authenticated if we send the "Authorization" header the request fails.

The magic string is just me reverse engineering others code, I could find any documentation on it. This authentication method is really nasty and poorly documented. That is just a base64 of "NTLMSSP00".

Reply:
btw searching for some other project that has handled this, I stumbled upon this https://github.com/Azure/go-ntlmssp

Have you investigated it? It seems like it's an open source MIT-licensed Go library from Microsoft that actually implements an NTLM `RoundTripper`!

Reply:
Yes I did. But by the time I tested, it didn't work correctly.

But I'm not doing anything NTLM related in this PR, any changes related to that should be done later in a different PR.

I'll just separate the code again to keep this at the VU level, otherwise, it will not work as supposed to.

Reply:
Seems like this library basically does what I hoped could be possible: https://github.com/Azure/go-ntlmssp/blob/master/negotiator.go#L52-L58

And maybe we can directly use it as a drop-in replacement, since it's just a `RoundTripper` implementation as well. So nothing new in `state`, no mysterious `cache` (or whatever we want to call it). Just have something like this:
```go
transport := netext.NewTransport(state, tags)
if preq.auth == "ntlm" {
  transport =  ntlmssp.Negotiator{
    RoundTripper: transport,
  }
}
```
And since the request-response tracking is in our transport right now, any NTLM negotiation requests would also be counted...

Reply:
Or It was me messing up. Which is probably the case 😂